spec/models/manageiq/providers/embedded_terraform/automation_manager/stack_spec.rb

```diff
@@ -30,4 +30,125 @@
+    let(:template) { FactoryBot.create(:terraform_template) }
+
+    context "when miq_task.job present" do
+      require 'webmock/rspec'
+      require 'json'
+      let(:terraform_runner_url) { "https://1.2.3.4:7000" }
+      let(:hello_world_retrieve_response) do
+        JSON.parse(File.read(File.join(__dir__, "../../../../../lib/terraform/runner/data/responses/hello-world-retrieve-success.json")))
+      end
+      let(:miq_task) do
+        miq_task = FactoryBot.create(:miq_task)
+        miq_task.job = job
+        miq_task
+      end
+
+      let(:job) do
+        ManageIQ::Providers::EmbeddedTerraform::AutomationManager::Job.create_job(template, {}, {}, []).tap do |job|
+          job.state = "finished"
+          job.options = {
+            :terraform_stack_id => hello_world_retrieve_response['stack_id']
+          }
+        end
+      end
+
+      let(:terraform_runner_stdout) { hello_world_retrieve_response['message'] }
+      let(:terraform_runner_stdout_html) { TerminalToHtml.render(terraform_runner_stdout) }
+
+      retrieve_stub = nil
+
+      before do
+        stack.miq_task = miq_task
+
+        stub_const("ENV", ENV.to_h.merge("TERRAFORM_RUNNER_URL" => terraform_runner_url))
+
+        retrieve_stub = stub_request(:post, "#{terraform_runner_url}/api/stack/retrieve")
+                        .with(:body => hash_including({:stack_id => hello_world_retrieve_response['stack_id']}))
+                        .to_return(
+                          :status => 200,
+                          :body   => hello_world_retrieve_response.to_json
+                        )
+      end
+
+      it "json" do
+        expect(stack.raw_stdout("json")).to eq terraform_runner_stdout
+
+        expect(retrieve_stub).to(have_been_requested.times(1))
+      end
+
+      it "txt" do
+        expect(stack.raw_stdout("txt")).to eq terraform_runner_stdout
+
+        expect(retrieve_stub).to(have_been_requested.times(1))
+      end
+
+      it "html" do
+        expect(stack.raw_stdout("html")).to eq terraform_runner_stdout_html
+
+        expect(retrieve_stub).to(have_been_requested.times(1))
+      end
+
+      it "nil" do
+        expect(stack.raw_stdout).to eq terraform_runner_stdout
+
+        expect(retrieve_stub).to(have_been_requested.times(1))
+      end
+    end
+
+    shared_examples_for "terraform runner stdout not available from miq_task" do
+      it "json" do
+        expect(stack.raw_stdout("json")).to eq("")
+      end
+
+      it "txt" do
+        expect(stack.raw_stdout("txt")).to eq ""
+      end
+
+      it "html" do
+        expect(stack.raw_stdout("html")).to include <<~EOHTML
+          <div class='term-container'>
+          No output available
+          </div>
+        EOHTML
+      end
+
+      it "nil" do
+        expect(stack.raw_stdout).to eq ""
+      end
+    end
+
+    context "when miq_task.job is missing" do
+      before do
+        stack.miq_task = nil
+      end
```

Comment:
I would do `let(:miq_task) { nil }` or override the stack definition like `let(:stack) { FactoryBot.create(:terraform_stack) }` rather than do this in a before block

Reply:
I guess, it is more readable this way,  
when looking at, what the test is doing

Reply:
`let(:miq_task) { nil }` is pretty clear what is happening, and you can get some weird behavior with before if you're not careful since they all run at each level it is easier to get messed up

Reply:
Do you mean in this way ?

```
    context "when miq_task is missing" do
      let(:miq_task) { nil }

      it_behaves_like "terraform runner stdout not available from miq_task"
    end

    context "when miq_task present, but missing miq_task.job" do
      let(:miq_task) { FactoryBot.create(:miq_task, :job => nil) }

      it_behaves_like "terraform runner stdout not available from miq_task"
    end

    context "when miq_task.job.options present but missing terraform_stack_id" do
      let(:job) do
        ManageIQ::Providers::EmbeddedTerraform::AutomationManager::Job.create_job(template, {}, {}, []).tap do |job|
          job.state = "waiting_to_start"
          job.options = {}
        end
      end

      let(:miq_task) { FactoryBot.create(:miq_task, :job => job) }

      it_behaves_like "terraform runner stdout not available from miq_task"
    end
```

if yes,  then 2nd & 3rd context, did not work as expected ...